// @algorandfoundation/algokit-client-generator/client/utility-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnCompleteCodeMap = exports.utilityTypes = void 0;
const writer_1 = require("../output/writer");
function* utilityTypes() {
    yield* (0, writer_1.jsDoc)(`Defines an onCompletionAction of 'no_op'`);
    yield `export type OnCompleteNoOp =  { onCompleteAction?: 'no_op' | OnApplicationComplete.NoOpOC }`;
    yield* (0, writer_1.jsDoc)(`Defines an onCompletionAction of 'opt_in'`);
    yield `export type OnCompleteOptIn =  { onCompleteAction: 'opt_in' | OnApplicationComplete.OptInOC }`;
    yield* (0, writer_1.jsDoc)(`Defines an onCompletionAction of 'close_out'`);
    yield `export type OnCompleteCloseOut =  { onCompleteAction: 'close_out' | OnApplicationComplete.CloseOutOC }`;
    yield* (0, writer_1.jsDoc)(`Defines an onCompletionAction of 'delete_application'`);
    yield `export type OnCompleteDelApp =  { onCompleteAction: 'delete_application' | OnApplicationComplete.DeleteApplicationOC }`;
    yield* (0, writer_1.jsDoc)(`Defines an onCompletionAction of 'update_application'`);
    yield `export type OnCompleteUpdApp =  { onCompleteAction: 'update_application' | OnApplicationComplete.UpdateApplicationOC }`;
    yield* (0, writer_1.jsDoc)('A state record containing a single unsigned integer');
    yield `export type IntegerState = {`;
    yield writer_1.IncIndent;
    yield* (0, writer_1.jsDoc)('Gets the state value as a BigInt ');
    yield `asBigInt(): bigint`;
    yield* (0, writer_1.jsDoc)('Gets the state value as a number.');
    yield `asNumber(): number`;
    yield writer_1.DecIndentAndCloseBlock;
    yield* (0, writer_1.jsDoc)('A state record containing binary data');
    yield `export type BinaryState = {`;
    yield writer_1.IncIndent;
    yield* (0, writer_1.jsDoc)('Gets the state value as a Uint8Array');
    yield `asByteArray(): Uint8Array`;
    yield* (0, writer_1.jsDoc)('Gets the state value as a string');
    yield `asString(): string`;
    yield writer_1.DecIndentAndCloseBlock;
}
exports.utilityTypes = utilityTypes;
exports.OnCompleteCodeMap = {
    no_op: 'OnCompleteNoOp',
    opt_in: 'OnCompleteOptIn',
    close_out: 'OnCompleteCloseOut',
    delete_application: 'OnCompleteDelApp',
    update_application: 'OnCompleteUpdApp',
};
